lib/publiccloud/instance.pm

```diff
@@ -349,56 +351,68 @@ sub wait_for_ssh {
     $args{ignore_wrong_pubkey} //= 0;
     $args{username} //= $self->username();
     $args{public_ip} //= $self->public_ip();
+    $args{systemup_check} //= 1;    # check also system up, defult:true; when false,systemup check not executed,only ssh checked.
     my $start_time = time();
-    my $check_port = 1;
-    my $sleep_period = $args{ignore_wrong_pubkey} ? 20 : 1;
-
-    # Looping until reaching timeout or passing two conditions :
-    # - SSH port 22 is reachable
-    # - journalctl got message about reaching one of certain targets
-    while ((my $duration = time() - $start_time) < $args{timeout}) {
-        if ($check_port) {
-            $check_port = 0 if (script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1) == 0);
-        }
-        else {
-            # On boottime test we do hard reboot which may change the instance address
-            script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
-
-            my $output = $self->run_ssh_command(
-                cmd => 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"',
-                proceed_on_failure => 1,
-                username => $args{username});
-            if ($output =~ m/Reached target.*/) {
-                return $duration;
-            }
-            elsif ($output =~ m/Permission denied \(publickey\).*/) {
-                die "ssh permission denied (pubkey)" unless $args{ignore_wrong_pubkey};
-            }
-        }
-        sleep $sleep_period;
+    my $instance_msg = sprintf("instance %s with public IP:%s within %d seconds", $self->{instance_id}, $self->{public_ip}, $args{timeout});
+
+    # Looping until SSH port 22 is reachable or timeout.
+    # Consider we that the args{timeout} is the Total timeout,
+    # while the script_retry's timeout is for each command retry,
+    # however we use the nc command with internal '-w 1' sec. timeout.
+    # So, we calculate the retry number of the check + delay, consuming the Total timeout,
+    # but with 10sec timeout as upper limit of the command run:
+    my $delay = $args{ignore_wrong_pubkey} ? 20 : 1;
+    my $status = script_retry('nc -vz -w 1 ' . $self->{public_ip} . ' 22',
+        timeout => 10, delay => $delay, retry => int($args{timeout} / ($delay + 1)) + 1,
+        fail_msg => "Unable to reach SSH port of $instance_msg", die => !$args{proceed_on_failure});
```

Comment:
```suggestion
    my $status = script_retry('nc -vz -w 1 ' . $self->{public_ip} . ' 22',
        timeout => 10, delay => 5, retry => int($args{timeout} / 5),
        fail_msg => "Unable to reach SSH port of $instance_msg", die => !$args{proceed_on_failure});
```

Reply:
`$args{ignore_wrong_pubkey}` does not have any effect in your suggestion

Reply:
I think that `delay 5` is too much, we use wait option `-w 1` in `nc` that already implements an internal loop, so instead better to increase it to `-w 5`.

I mantained some settings of the old code, as in master, but not clear utility of `ignore_wrong_pubkey` opt., having already `proceed_on_failure`, 
neither clear why *delay* (old name sleep_period) should be so big 20s when set  true.